src/abstract_block.jl

```diff
@@ -230,10 +235,11 @@ Dispatch parameters in collection to block tree `x`.
 end
 
 @interface function dispatch!(f::Union{Function, Nothing}, x::AbstractBlock, it)
+    @assert (it isa Symbol || length(it) == nparameters(x)) "expect $(nparameters(x)) parameters, got $(length(it))"
```

Comment:
`dp.loc` should be same as `nparameters(x)`, except it does not make this algorithm complexity `O(N^2)`.
So, what is reason to change this line? @Roger-luo

Reply:
I don't get it, `dp.doc` is `0` here, isn't it?

Reply:
`Dispatcher(it)` calls `Dispatcher(it, 0)`

Reply:
This code checks the total length of parameters, instead of checking the full length.

`consume!` increase `loc`, at the end of dispatch, we check if `loc` (the number of consumed parameters) is consistent with the length of parameters.

Reply:
This is causing the following behaviour:

```julia
julia> g = chain(Rx(0.1), Ry(0.2))
nqubits: 1, datatype: Complex{Float64}
chain
├─ rot(X gate, 0.1)
└─ rot(Y gate, 0.2)

julia> dispatch!(g, (0.2, ))
ERROR: BoundsError: attempt to access (0.2,)
  at index [2]
Stacktrace:
 [1] getindex at ./tuple.jl:24 [inlined]
 [2] getindex at ./range.jl:292 [inlined]
 [3] consume! at /Users/roger/.julia/dev/YaoBlocks/src/abstract_block.jl:211 [inlined]
 [4] dispatch!(::Nothing, ::RotationGate{1,Float64,YGate{Complex{Float64}}}, ::YaoBlocks.Dispatcher{Tuple{Float64}}) at /Users/roger/.julia/dev/YaoBlocks/src/abstract_block.jl:225
 [5] dispatch!(::Nothing, ::ChainBlock{1,Complex{Float64}}, ::YaoBlocks.Dispatcher{Tuple{Float64}}) at /Users/roger/.julia/dev/YaoBlocks/src/abstract_block.jl:227
 [6] dispatch!(::Nothing, ::ChainBlock{1,Complex{Float64}}, ::Tuple{Float64}) at /Users/roger/.julia/dev/YaoBlocks/src/abstract_block.jl:234
 [7] dispatch!(::ChainBlock{1,Complex{Float64}}, ::Tuple{Float64}) at /Users/roger/.julia/dev/YaoBlocks/src/abstract_block.jl:239
 [8] top-level scope at none:0

julia> g
nqubits: 1, datatype: Complex{Float64}
chain
├─ rot(X gate, 0.2)
└─ rot(Y gate, 0.2)
```

Reply:
I think we shouldn't change the parameters, if the input is not correct.